File: scripts/40_q3_var_recursive_forecast.R
# scripts/40_q3_var_recursive_forecast.R
# HW2 Section 3: one-step-ahead recursive forecasts (expanding window) + RMSE

suppressPackageStartupMessages({
  library(readxl)
  library(vars)
})

source("R/q1_io.R")
source("R/q23_var_core.R")

date_col <- "date"
y_col <- "import_clv_qna_sa"

x_path <- "data/raw/x.xlsx"
y_path <- "data/raw/y.xlsx"

# Read data
x_df <- read_x_strict_one_sheet(x_path)
y_df <- read_y_data_y_sheet(y_path, sheet = "data_y")

# Get selected x from Q1 output
top5 <- read.csv("output/tables/q1_top5_abs_corr_R.csv", stringsAsFactors = FALSE)
x_code <- top5$code[1]
if (!(x_code %in% names(x_df))) stop(sprintf("Selected x_code '%s' not in x.xlsx.", x_code), call. = FALSE)

# Load chosen lag p*
p_star <- as.integer(trimws(readLines("output/tables/q2_var_selected_lag.txt")))
if (!is.finite(p_star) || p_star < 1) stop("Invalid p* in output/tables/q2_var_selected_lag.txt", call. = FALSE)

panel <- prepare_var_levels(
  y_df = y_df,
  x_df = x_df,
  date_col = date_col,
  y_col = y_col,
  x_col = x_code,
  allow_missing_last_x = TRUE
)

n <- nrow(panel)
n_train <- q23_train_end(n, 0.8)

# build full differenced data
diff_df <- q23_first_difference(panel, date_col, cols = c(y_col, x_code), diff_lag = 1)
mat_all <- diff_df[, c(paste0("d_", y_col), paste0("d_", x_code)), drop = FALSE]
colnames(mat_all) <- c("dy", "dx")

# recursive forecasts for level y
y_level <- panel[[y_col]]
dates <- panel[[date_col]]

targets <- (n_train + 1):n

pred_list <- lapply(targets, function(t) {
  # use diffs up to (t-1) -> in diff space row (t-1)-1 = t-2
  end_diff <- t - 2
  train_mat <- mat_all[1:end_diff, , drop = FALSE]

  fit <- vars::VAR(train_mat, p = p_star, type = "const")
  fc <- predict(fit, n.ahead = 1)$fcst$dy[1, "fcst"]
  y_hat <- y_level[t - 1] + as.numeric(fc)

  data.frame(
    date = dates[t],
    y_true = y_level[t],
    y_hat = y_hat
  )
})

pred <- do.call(rbind, pred_list)
rmse <- q23_rmse(pred$y_true, pred$y_hat)

dir.create("output/tables", recursive = TRUE, showWarnings = FALSE)
utils::write.csv(pred, "output/tables/q3_var_recursive_forecasts.csv", row.names = FALSE)
utils::write.csv(data.frame(rmse = rmse), "output/tables/q3_var_rmse.csv", row.names = FALSE)

message(sprintf("VAR recursive RMSE (level y): %.6f", rmse))
message("Wrote: output/tables/q3_var_recursive_forecasts.csv")
message("Wrote: output/tables/q3_var_rmse.csv")

File: scripts/40_q3_var_recursive_forecast.py
# scripts/40_q3_var_recursive_forecast.py
from __future__ import annotations

from pathlib import Path
import sys

PROJECT_ROOT = Path(__file__).resolve().parents[1]
if str(PROJECT_ROOT) not in sys.path:
    sys.path.insert(0, str(PROJECT_ROOT))

import pandas as pd

from python.var_core import (
    prepare_var_level_data,
    recursive_var_one_step_forecast_level,
    compute_rmse,
)
from python.var_io import (
    read_x_xlsx,
    read_y_xlsx,
    read_top1_x_code_from_q1_csv,
    read_txt_int,
    write_csv,
    write_txt,
)


def main() -> None:
    date_col = "date"
    y_col = "import_clv_qna_sa"

    x_path = "data/raw/x.xlsx"
    y_path = "data/raw/y.xlsx"

    q1_csv = "output/tables/q1_top5_abs_corr_R.csv"
    x_code = read_top1_x_code_from_q1_csv(q1_csv)

    x_df = read_x_xlsx(x_path, sheet_name="data_x")
    y_df = read_y_xlsx(y_path, sheet_name="data_y")

    if x_code not in x_df.columns:
        raise ValueError(f"Selected x_code '{x_code}' not found in x.xlsx (data_x).")

    panel = prepare_var_level_data(
        y_df=y_df,
        x_df=x_df,
        date_col=date_col,
        y_col=y_col,
        x_col=x_code,
        allow_last_x_missing=True,
    )

    # Prefer Python-selected lag; fallback to R if needed
    py_lag_path = Path("output/tables_py/q2_var_selected_lag_py.txt")
    if py_lag_path.exists():
        p_star = read_txt_int(str(py_lag_path))
    else:
        p_star = read_txt_int("output/tables/q2_var_selected_lag.txt")

    pred = recursive_var_one_step_forecast_level(
        level_df=panel,
        date_col=date_col,
        y_col=y_col,
        x_col=x_code,
        p=p_star,
        train_ratio=0.8,
        diff_lag=1,
        trend="c",
    )

    rmse_val = compute_rmse(pred["y_true"].to_numpy(), pred["y_pred"].to_numpy())
    write_csv(pred, "output/tables_py/q3_var_recursive_forecasts_py.csv")
    write_txt(f"{rmse_val:.6f}", "output/tables_py/q3_var_rmse_py.txt")

    # Optional consistency check with R (do not hard fail on tiny float diffs)
    r_rmse_path = Path("output/tables/q3_var_rmse.csv")
    if r_rmse_path.exists():
        r_rmse = float(pd.read_csv(r_rmse_path)["rmse"].iloc[0])
        if abs(r_rmse - rmse_val) > 1e-3:
            raise RuntimeError(f"Python RMSE={rmse_val} differs from R RMSE={r_rmse} (>1e-3).")

    print(f"Section3 (Python) done. VAR recursive RMSE = {rmse_val:.6f}")


if __name__ == "__main__":
    main()
